proposals/shared-everything-threads/Overview.md

```diff
@@ -751,6 +751,42 @@ still must ensure that non-atomic accesses to any reference do not tear.
 > large sizes? It may be the case that strategies to prevent tearing also admit synchronizing
 > accesses.
 
+### Text format
+
+The text format must change to accommodate `shared` attributes in many places. In general, the
+`shared` attribute _prefixes_ any existing modifiers, such as with global types, e.g.:
+
+```
+globaltype  ::=  t:valtype
+              |  '(' 'mut' t:valtype ')'
+              |  '(' 'shared' t:valtype ')'
+              |  '(' 'shared' 'mut' t:valtype ')'
+```
+
+Both the short and long form of heap types accept a `shared` prefix, albeit with extra parentheses.
+E.g.:
+
+```
+'anyref'                   =  '(' 'ref' 'null' 'any' ')'
+'(' 'shared' 'anyref' ')'  =  '(' 'ref' 'null' '(' 'shared' 'any' ')' ')'
+...
+```
+
+Though only the `any` heap type is shown here, this is applicable to all other abstract heap types.
+It /does not/ apply to concrete heap types (e.g., `(ref null $t)`) since the `shared`-ness will be
+attached to the pointed-to type (e.g., `$t`); thus, `(ref null (shared $t))` is invalid text.
+
+Likewise, composite types can be `shared`, but here we /wrap/ instead of /prefix/:
+
+```
+sharecomptype  ::=  'shared' '(' at:arraytype ')'
+                 |  'shared' '(' st:structtype ')'
+                 |  'shared' '(' ft:functype ')'
+```
+
+What this means is that, to define the type of a shared function with no parameters or results, one
+would write: `(type (shared (func)))`.
+
```

Comment:
When I got here I was still debating whether to describe an alternate format for marking globals, tables, and memories as `shared` based on the previous paragraph. Currently, memories are marked `shared` by suffixing the `shared` attribute on the end (e.g., `(memory 1 1 shared)`); this has the advantage that, in a parser, one sees the `shared` token within the context of parsing the memory type. And this is how I spec-ed globals above: e.g., `(global (shared i32) (i32.const 0))` &mdash; `shared` is /within/ the type.

But if we start wrapping composite types up with `(shared (...))` as described above, then there's been [discussion](https://github.com/WebAssembly/shared-everything-threads/issues/61) suggesting that:

(a) globals, tables, and memories gain the same "wrapping" syntax, e.g., `(shared (memory 1 1))`, `(shared (global i32 (i32.const 0))`, etc.
 
(b) we move the `shared` attribute on memories to be the first attribute, e.g., `(memory shared 1 1)`, like I planned to do for globals

The problem with both of these is that we might have to maintain both alternatives for a while. (a) also has the annoying disadvantage in that it requires more invasive changes to current parsers. In any case, my feeling is that we don't want three different kinds of labeling things `shared`: /prefixes/, /suffixes/, /wrapping/.

Should we just always /suffix/?